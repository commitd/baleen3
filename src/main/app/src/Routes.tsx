/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { Router } from '@reach/router'
import React from 'react'
import {
  Components,
  Help,
  NotFound,
  PipelineEdit,
  PipelineNew,
  Home,
  PipelineView,
} from './pages'

export const Routes: React.FC = () => (
  <Router>
    <NotFound default />
    <Home path="/" />
    <PipelineNew path="new" />
    <PipelineView path="view/:pipelineId" />
    <PipelineEdit path="edit/:pipelineId" />
    <Components path="components" />
    <Help path="help" />
  </Router>
)
