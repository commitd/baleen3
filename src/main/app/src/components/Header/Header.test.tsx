/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { Icons } from '@committed/components'
import React from 'react'
import { Routes } from '../../Routes'
import {
  render,
  renderDark,
  renderWithRouter,
  userEvent,
} from '../../utils/test'
import { Header, MainAction, ToolbarAction, ToolbarSelect } from '.'

// NB
// We don't snapshot some of these as we found that sometimes the snapshots have a touch ripple element and sometimes they don't.
// I think this is a timing issue, but it meant some tests failed occasionally.
// Instead we just check there is some output rather than have sporadically failing tests.

jest.mock('../../containers/ComponentsContainer', () => {
  const FakePageContainer = jest.fn(() => null)
  return { ComponentsContainer: FakePageContainer }
})

jest.mock('../../containers/PipelinesContainer', () => {
  const FakePageContainer = jest.fn(() => null)
  return { PipelinesContainer: FakePageContainer }
})

it('Header renders without error', () => {
  const { asFragment } = render(<Header />)
  expect(asFragment()).toMatchSnapshot()
})

it('Header renders dark without error', () => {
  const { asFragment } = renderDark(<Header />)
  expect(asFragment()).toMatchSnapshot()
})

it('handles navigation clicks', async () => {
  //NB We use Routes here to have the navigation, but we are really testing the embedded headers within the pages.
  // To check the navigation links function
  const { findByTestId, getByRole, getAllByRole } = renderWithRouter(<Routes />)

  expect(await findByTestId('Home')).toBeDefined()

  userEvent.click(getByRole('link', { name: /components/i }))
  expect(await findByTestId('Components')).toBeDefined()

  userEvent.click(getByRole('link', { name: /help/i }))
  expect(await findByTestId('Help')).toBeDefined()

  userEvent.click(getAllByRole('link', { name: /Baleen/i })[0])
  expect(await findByTestId('Home')).toBeDefined()

  userEvent.click(getByRole('button', { name: /components/i }))
  expect(await findByTestId('Components')).toBeDefined()

  userEvent.click(getByRole('button', { name: /help/i }))
  expect(await findByTestId('Help')).toBeDefined()

  userEvent.click(getAllByRole('button', { name: /Baleen/i })[0])
  expect(await findByTestId('Home')).toBeDefined()
}, 30000)

it('MainAction renders without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <MainAction onClick={onClick}>Action</MainAction>
  )
  expect(asFragment()).toMatchSnapshot()
})

it('MainAction renders dark without error', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <MainAction onClick={onClick}>Action</MainAction>
  )
  expect(asFragment()).toBeDefined()
})

it('MainAction renders with icon without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <MainAction onClick={onClick} icon={<Icons.Star />}>
      Action
    </MainAction>
  )
  expect(asFragment()).toBeDefined()
})

it('MainAction onClick is called', () => {
  const onClick = jest.fn()
  const { getByRole } = render(
    <MainAction onClick={onClick} icon={<Icons.Star />}>
      Action
    </MainAction>
  )
  expect(onClick).not.toBeCalled()
  userEvent.click(getByRole('button', { name: /Action/i }))
  expect(onClick).toBeDefined()
})

it('ToolbarAction renders without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <ToolbarAction onClick={onClick}>Action</ToolbarAction>
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarAction renders dark without error', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <ToolbarAction onClick={onClick}>Action</ToolbarAction>
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarAction renders with icon', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <ToolbarAction onClick={onClick} icon={<Icons.Save />}>
      Action
    </ToolbarAction>
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarAction click calls callback', () => {
  const onClick = jest.fn()
  const { getByRole } = renderDark(
    <ToolbarAction onClick={onClick} icon={<Icons.Save />}>
      Action
    </ToolbarAction>
  )

  expect(onClick).not.toBeCalled()
  userEvent.click(getByRole('button', { name: /Action/i }))
  expect(onClick).toHaveBeenCalled()
})

it('ToolbarSelect renders without error', () => {
  const onClick = jest.fn()
  const { asFragment } = render(
    <ToolbarSelect
      selectedIndex={0}
      onClick={onClick}
      options={['Option 1', 'Option 2']}
    />
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarSelect renders dark without error', () => {
  const onClick = jest.fn()
  const { asFragment } = renderDark(
    <ToolbarSelect
      selectedIndex={0}
      onClick={onClick}
      options={['Option 1', 'Option 2']}
    />
  )
  expect(asFragment()).toBeDefined()
})

it('ToolbarSelect renders placeholder', () => {
  const onClick = jest.fn()
  const { getByText } = render(
    <ToolbarSelect
      selectedIndex={-1}
      onClick={onClick}
      placeholder="Test"
      options={['Option 1', 'Option 2']}
    />
  )
  getByText('Test')
})

it('ToolbarSelect shows selected option', () => {
  const onClick = jest.fn()
  const { getByText } = render(
    <ToolbarSelect
      selectedIndex={1}
      onClick={onClick}
      placeholder="Test"
      options={['Option 1', 'Option 2']}
    />
  )
  getByText('Option 2')
})

it('ToolbarSelect can select another option', () => {
  const onClick = jest.fn()
  const options = ['Option 1', 'Option 2']
  const { getByRole, getByText } = render(
    <ToolbarSelect selectedIndex={0} onClick={onClick} options={options} />
  )

  expect(onClick).not.toBeCalled()
  userEvent.click(getByRole('button', { name: options[0] }))
  expect(onClick).not.toBeCalled()
  userEvent.click(getByText(options[1]))
  expect(onClick).toHaveBeenCalledWith(1)
})
