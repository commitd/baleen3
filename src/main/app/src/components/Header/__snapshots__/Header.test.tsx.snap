// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders dark without error 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary WithStyles(ForwardRef(AppBar))-root WithStyles(ForwardRef(AppBar))-root Styled(WithStyles(ForwardRef(AppBar)))-root Styled(WithStyles(ForwardRef(AppBar)))-root MuiPaper-elevation0"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular Component-root Component-root MuiToolbar-gutters"
        >
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          >
            <a
              aria-current="page"
              class="Link-root"
              href="/"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    class="ForwardRef(SvgBaleenHorizontal)-root ForwardRef(SvgBaleenHorizontal)-root"
                  >
                    Baleen_Horizontal.svg
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </a>
          </div>
          <a
            aria-label="components"
            class="Link-root"
            href="/components"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root"
              tabindex="0"
              title="Components"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12.09 2.91C10.08.9 7.07.49 4.65 1.67l4.34 4.34-3 3-4.34-4.34C.48 7.1.89 10.09 2.9 12.1c1.86 1.86 4.58 2.35 6.89 1.48l9.82 9.82 3.71-3.71-9.78-9.79c.92-2.34.44-5.1-1.45-6.99z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </a>
          <a
            aria-label="help"
            class="Link-root"
            href="/help"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root"
              tabindex="0"
              title="Help"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </a>
          <button
            aria-label="switch-theme"
            class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root MuiIconButton-colorPrimary"
            tabindex="0"
            title="Use dark theme"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-color makeStyles-color"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 2c-1.05 0-2.05.16-3 .46 4.06 1.27 7 5.06 7 9.54s-2.94 8.27-7 9.54c.95.3 1.95.46 3 .46 5.52 0 10-4.48 10-10S14.52 2 9 2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;

exports[`Header renders without error 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary WithStyles(ForwardRef(AppBar))-root WithStyles(ForwardRef(AppBar))-root Styled(WithStyles(ForwardRef(AppBar)))-root Styled(WithStyles(ForwardRef(AppBar)))-root MuiPaper-elevation0"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular Component-root Component-root MuiToolbar-gutters"
        >
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          >
            <a
              aria-current="page"
              class="Link-root"
              href="/"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    class="ForwardRef(SvgBaleenHorizontal)-root ForwardRef(SvgBaleenHorizontal)-root"
                  >
                    Baleen_Horizontal.svg
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </a>
          </div>
          <a
            aria-label="components"
            class="Link-root"
            href="/components"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root"
              tabindex="0"
              title="Components"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12.09 2.91C10.08.9 7.07.49 4.65 1.67l4.34 4.34-3 3-4.34-4.34C.48 7.1.89 10.09 2.9 12.1c1.86 1.86 4.58 2.35 6.89 1.48l9.82 9.82 3.71-3.71-9.78-9.79c.92-2.34.44-5.1-1.45-6.99z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </a>
          <a
            aria-label="help"
            class="Link-root"
            href="/help"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root"
              tabindex="0"
              title="Help"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </a>
          <button
            aria-label="switch-theme"
            class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root MuiIconButton-colorPrimary"
            tabindex="0"
            title="Use dark theme"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-color makeStyles-color"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 2c-1.05 0-2.05.16-3 .46 4.06 1.27 7 5.06 7 9.54s-2.94 8.27-7 9.54c.95.3 1.95.46 3 .46 5.52 0 10-4.48 10-10S14.52 2 9 2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;

exports[`MainAction renders without error 1`] = `
<DocumentFragment>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained WithNoOpener-root WithNoOpener-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Action
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
